### DISCUSSION TITLE
Fehler beim Hochladen der Firmware Image in den RAM

### DISCUSSION THREAD

**Discussion by fskander** [PeterPawn/YourFritz#40] [Q&A]
Hallo,

Erstmal Entschuldigung im Voraus wenn ich Fehler mache, Deutsch ist für mich eine Fremdsprache ;)

Ich möchte Freetz auf meinem Fritzbox 7530 installieren, u.a um OpenVPN und/oder WireGuard nutzen zu können.
Ich könnte die Freetz Image [mithilfe dieser Anleitung](https://freetz-ng.github.io/freetz-ng/wiki/10_Beginner/newbie.de.html ) erstellen, ich habe aber Probleme wenn es kommt zu flashen.
Die angegebene Methode (Upload uber die AVM Interface) funktioniert nicht mehr, ich habe ein paar Recherche gemacht und ich bin auf [diese Anleitung](https://web.archive.org/web/20190529205946/https://www.ip-forum.eu/howto-fritz-box-firmware-images-auch-unsignierte-ueber-den-bootloader-installieren-577) aufmerksam geworden.

- Ich bin auf Windows10 und ich habe PowerShell v7.1 heruntergeladen
- Der Bootloader könnte ich mit der 1. Methode (Ethernet) erreichen und könnte mich auch einloggen
- Firmware Image erstellen/vorbereiten : OK
- Der Fehler kommt wenn ich versuche die Firmware Image in den RAM hochzuladen. Wenn ich die folgende Kommandozeile in PowerShell eingebe :
```
c:\YourFritz\master\eva_tools\EVA-FTP-Client.ps1 -Verbose -Debug -ScriptBlock { BootDeviceFromImage c:\YourFritz\Images\firmware.image.in-memory }
0
```
Bekomme ich dann die folgende Fehlermeldung :
```
"Exception calling ".ctor" with "2 argument(s): "No connection could be made because the target machine actively refused it. 192.168.178.1:21".
Error connecting to remote host
```
<img width="1280" alt="2021-07-04 (1)" src="https://user-images.githubusercontent.com/86897851/124396428-c84f0c80-dd09-11eb-9ee8-920204f6c5ed.png">

Ich habe mein Firewall (Windows Defender) deaktiviert, es hat sich nichts geändert. Was mache ich falsch ?

**Comment by PeterPawn** [comment]
Ich weiß nicht, wer das Bild tatsächlich lesen kann ... ich kann es nicht. Die Auflösung ist miserabel - bei mir gibt es keine Javascript-Freigabe für derartige Sites, zumal man die Bilder hier auch direkt anhängen kann.

Das, was man da "erahnen" kann beim Zusammenstellen der Datei für den Upload sieht aber auch falsch aus - ich war mir nicht bewußt, daß die Anleitung von @qwertzasdfgh solchen Unsinn enthält. Für den Start aus dem RAM wird ein Kernel und ein Dateisystem benötigt (und zwar als EINE Datei) und das kann natürlich nicht nur mit dem Extrahieren eines Kernel-Images erstellt werden.

Das korrekte Erstellen eines bootbaren Images aus einem TAR-File (egal ob von AVM oder Freetz[-NG]) ist hier: https://github.com/PeterPawn/YourFritz/tree/main/signimage dokumentiert - denn da steht auch das Skript-File (weil es mehr mit Signieren als Extrahieren zu tun hat, das sind nur "Nebenfunktionen"). Es gibt Alternativen, das mit mehreren Aufrufen und Dateien auch selbst zu erstellen - die betreffende Funktion ist auch nichts anderes als eine Zusammenfassung mehrerer anderer: https://github.com/PeterPawn/YourFritz/blob/90d8bcb42d4a85702bc211a57d44d60b19387370/signimage/FirmwareImage.ps1#L2595

---

Auch bezweifele ich, daß es machbar ist, zwei parallele FTP-Verbindungen zum FTP-Server in EVA aufzubauen - das Fenster oben links sieht nach einer "verwaisten" FTP-Verbindung aus, die nicht geschlossen wurde und parallel dazu scheint aus der PS-Console heraus ein weiterer Zugriff (mit `GetEnvironmentValue`) versucht zu werden.

Egal, ob Du das nun mit meinen Skripts oder mit der von Dir gewünschten Alternative realisieren willst - solange da zwei FTP-Sessions versucht werden, wird das mit einiger Wahrscheinlichkeit scheitern ... ich wüßte momentan keine EVA-Version, die parallele Zugriffe erlaubt (zumal das ja echt gefährlich wird, wenn da gerade Flash gelöscht wurde und neu beschrieben werden soll, während parallel ein `REBOOT` hereinkommt).

---

Wenn es Dir angenehmer ist, in Englisch zu kommunizieren, ist das tatsächlich kein Problem - ich habe nur auf Deutsch geantwortet, weil die Frage auch in dieser Sprache gestellt war. Mein Französisch wäre aber zu schlecht ...

**Reply by fskander** [reply]
Thank you for the answer - I can read and understand German but English is indeed easier for me. :) I didn't know it was possible to upload images directly, I just edited that so it should be now readable.

So I understand the problem is that I'm using two different windows (PowerShell and Cmd Prompt) in parallel ? Should I connect to the Bootloader in PowerShell too ? Is there a correct, up to date step-by-step guide (in German or in English) available somewhere that I could use ? I'm not really familiar with all of this (running scripts with command lines, etc)

Thanks !

**Comment by PeterPawn** [comment]
Annotations/suggestions by me:

1. Create your image file first - I think, the linked text (https://github.com/PeterPawn/YourFritz/tree/main/signimage - last example) should be sufficient and all other preparations have been done already, according to your picture above (and by the way ... thanks for your changes to the first post).

2. Detect/discover the starting FRITZ!Box device using the `EVA-Discover.ps1` script. Power off the device, start the script `./EVA-Discover.ps1` (at your option with `-Verbose` and/or `-Debug` parameters, too) and power on the device again (as you've done already with your FTP attempt in the picture). You should not need additional parameters, but you may specify a `requested_address` for the device, matching your own network configuration. It's not really necessary to use IP address 192.168.178.1 - it's only the default value. The address of the found device will be logged as `EVA_IP=<something>` to the PS host at the end. 
 
The device (if any was found) was hold up in the "EVA state" already and there's no further need to use a FTP client or any other (foreign) software. The FRITZ!Box router will not continue its usual work, until it:

- was powered off and on again
- got a `REBOOT` command via a FTP session
- got an image (with proper format) loaded to RAM ... in this case the image gets started immediately after the upload command completed

3. Use `EVA-FTP-Client.ps1` to upload the image from step 1 to the device. Your documented attempt from above looks good at first glance (incl. the additional `0` for the `limitedMemory` option, even if `$False` would look more logical) - you fell by the second FTP connection only, as far as I can see.

4. The value of `linux_fs_start` only rules, which partition set will be used by the uploaded image to make itself persistent by writing itself to flash storage. If you want to preserve the latest running system, you have to switch the content of `linux_fs_start` (from `0` to `1` or vice versa) PRIOR TO the upload call - but you may simply use the `SwitchSystem` function from `EVA-FTP-Client.ps1`, too. If you want to run more than one single function call within a single script call, chain the functions together with a semicolon (`... -ScriptBlock { SwitchSystem; BootDeviceFromImage ... }`)

5. If you run into problems, provide the logs please - simply copy the text from the PS host window into a code block here (a code block is opened and closed by three consecutive "back-tick" characters on a single line).

**Reply by fskander** [reply]
It worked, thanks a lot !!

I was close - as you guessed, the issue was indeed that I was using two different windows in parallel for the FTP connection. Also, the first time I tried to discover the Fritzbox with the EVA-Discover method I couldn't find it...which was because I was not connected to the Fritzbox in Ethernet but by WiFi, which obviously cuts the connection between the PC and the Fritzbox when the Fritzbox reboots...yes I know it sounds very dumb and obvious, but it took me a while and several failed attempts to realize that.

By connecting in Ethernet, running both EVA-Discover and EVA-FTP-Client in Powershell, it worked like a charm. I did not have to recreate the image BTW - it worked with the method described in the post from @qwertzasdfgh : 
```
. c:\YourFritz\master\signimage\FirmwareImage.ps1
[FirmwareImage]::new("c:\YourFritz\Images\firmware.image").getBootableImage("c:\YourFritz\Images\firmware.image.in-memory")
```
